### PR TITLE
Ghost Roles Patch And Some changes

### PR DESCRIPTION
**Pull request by Ultradragon005**
- [x] Ghostassign directly in RpcSetRole, fix some bugs here.
- [x] Make ghost-roles compatible.
- [x] Make RoleClass[CustomRole] to use general roles and not only existing roles.
- [x] Rework Roles:
      * Hawk - Check Lang_US
      * BloodMoon - Check Lang_US

- [x] Add message roles has changed if "ShowRoleInfoFirstMeeting" Enabled.
- [x] Add More Options
      * "Converted Player Can Become Any GhostRole"
      * "Max Crew GhostRoles"
      * "Max Imp GhostRoles"

 - [x] Other changes: fix onmeetinghudstart, make sure to add "using UnityEngine.GraphicsBuffer;" everywhere for smoother performance ;)

### CODE REVIEW

**Comment by Ultradragon005** [comment]
goddamit I forgot to set it to rolebase again

File: Patches/PlayerControlPatch.cs [hunk 1]
```diff
@@ -1620,18 +1624,21 @@ public static bool Prefix(PlayerControl __instance, [HarmonyArgument(0)] ref Rol
...
                     roleType = RoleTypes.ImpostorGhost;
+                    return true;
                 }
                 else
                 {
-                    foreach ((var seer, var role) in ghostRoles)
+                    foreach ((var seer, var role) in ghostRoles) // Idk wtf this for tbh
                     {
                         Logger.Info($"Desync {targetName} => {role} for {seer.GetNameWithRole().RemoveHtmlTags()}", "PlayerControl.RpcSetRole");
                         target.RpcSetRoleDesync(role, seer.GetClientId());
```
Comment (Tommy-XL) [inline Patches/PlayerControlPatch.cs:1644]: This code assigns the roles of ghosts desync, that is, if the role is non-impostor, this player will assign ImpostorGhost, but for others players it's will be assigned as CrewmateGhost, if you test this in the game and look at the logs, then everything will become clear

**Comment by Ultradragon005** [comment]
image
OH HELL NAWH

**Comment by Ultradragon005** [comment]
okay we gud now

File: Patches/PlayerControlPatch.cs [hunk 2]
```diff
@@ -1583,6 +1566,58 @@ public static bool Prefix(PlayerControl __instance, [HarmonyArgument(0)] ref Rol
...
+                    writer.StartRpc(host.NetId, (byte)RpcCalls.SendChat)
+                        .Write(sb.ToString())
+                        .EndRpc();
+                    writer.StartRpc(host.NetId, (byte)RpcCalls.SetName)
+                        .Write(name)
+                        .EndRpc();
+                    writer.EndMessage();
+                    writer.SendMessage();
+                    host.Notify("", 0.1f);
+
```
Comment (Tommy-XL) [inline Patches/PlayerControlPatch.cs:1615]: Why jast not use `Utils.SendMessage()`?

Comment (Ultradragon005) [reply Patches/PlayerControlPatch.cs:1615]: mk, but I wanted it to send the moment player dies

Comment (Tommy-XL) [reply Patches/PlayerControlPatch.cs:1615]: `Utils.SendMessage()` сan send a message at any time, just call it

File: Roles/Neutral/Solsticer.cs [hunk 3]
```diff
@@ -317,7 +317,7 @@ public override void OnReportDeadBody(PlayerControl reporter, PlayerControl targ
     {
         patched = false;
     }
-    public override void OnMeetingHudStart(PlayerControl pc)
+    public override void OnOthersMeetingHudStart(PlayerControl pc)
```
Comment (Tommy-XL) [inline Roles/Neutral/Solsticer.cs:320]: Same here, the code is calculated only for Solsticer

File: Roles/Neutral/God.cs [hunk 4]
```diff
@@ -46,7 +46,7 @@ public override bool GuessCheck(bool isUI, PlayerControl guesser, PlayerControl
         return false;
     }
 
-    public override void OnMeetingHudStart(PlayerControl pc)
+    public override void OnOthersMeetingHudStart(PlayerControl pc)
```
Comment (Tommy-XL) [inline Roles/Neutral/God.cs:49]: Same here, the code is designed only for God and the message is sent to everyone if necessary

File: Roles/Crewmate/Retributionist.cs [hunk 5]
```diff
@@ -216,7 +216,7 @@ private static void RetributionistOnClick(byte playerId /*, MeetingHud __instanc
         if (AmongUsClient.Instance.AmHost) RetributionistMsgCheck(PlayerControl.LocalPlayer, $"/ret {playerId}", true);
         else SendRPC(playerId);
     }
-    public override void OnMeetingHudStart(PlayerControl pc)
+    public override void OnOthersMeetingHudStart(PlayerControl pc)
```
Comment (Tommy-XL) [inline Roles/Crewmate/Retributionist.cs:219]: Same here

File: Roles/Crewmate/Mortician.cs [hunk 6]
```diff
@@ -123,7 +123,7 @@ public override string GetSuffix(PlayerControl seer, PlayerControl target = null
         }
         else return "";
     }
-    public override void OnMeetingHudStart(PlayerControl pc)
+    public override void OnOthersMeetingHudStart(PlayerControl pc)
     {
         if (msgToSend.ContainsKey(pc.PlayerId))
```
Comment (Tommy-XL) [inline Roles/Crewmate/Mortician.cs:128]: Same here

File: Roles/Crewmate/Enigma.cs [hunk 7]
```diff
@@ -151,7 +151,7 @@ public override void OnReportDeadBody(PlayerControl player, PlayerControl target
         }
     }
 
-    public override void OnMeetingHudStart(PlayerControl pc)
+    public override void OnOthersMeetingHudStart(PlayerControl pc)
     {
         if (MsgToSend.ContainsKey(pc.PlayerId))
             AddMsg(MsgToSend[pc.PlayerId], pc.PlayerId, Utils.ColorString(Utils.GetRoleColor(CustomRoles.Enigma), MsgToSendTitle[pc.PlayerId]));
```
Comment (Tommy-XL) [inline Roles/Crewmate/Enigma.cs:157]: Same

**Pull request merged**